esphome/components/graph/graph.h

```diff
@@ -0,0 +1,181 @@
+  int xs{0};  // X spacing between labels
+  int ys{0};  // Y spacing between labels
+  int yl{0};  // Y spacing from label to line sample
+  int xv{0};  // X distance between label to value text
+  int yv{0};  // Y distance between label to value text
+  friend Graph;
+};
+
+class HistoryData {
+ public:
+  void init(int length);
+  ~HistoryData();
+  void set_update_time_ms(uint32_t update_time_ms) { update_time_ = update_time_ms; }
+  void take_sample(float data);
+  int get_length() const { return length_; }
+  float get_value(int idx) const { return samples_[(count_ + length_ - 1 - idx) % length_]; }
+  float get_recent_max() const { return recent_max_; }
+  float get_recent_min() const { return recent_min_; }
+
+ protected:
+  uint32_t last_sample_;
+  uint32_t period_{0};       /// in ms
+  uint32_t update_time_{0};  /// in ms
+  int length_;
+  int count_{0};
+  float recent_min_{NAN};
+  float recent_max_{NAN};
+  std::vector<float> samples_;
+};
+
+class GraphTrace {
+ public:
+  void init(Graph *g);
+  void set_name(std::string name) { name_ = name; }
+  void set_sensor(sensor::Sensor *sensor) { sensor_ = sensor; }
+  uint8_t get_line_thickness() { return this->line_thickness_; }
+  void set_line_thickness(uint8_t val) { this->line_thickness_ = val; }
+  enum LineType get_line_type() { return this->line_type_; }
+  void set_line_type(enum LineType val) { this->line_type_ = val; }
+  Color get_line_color() { return this->line_color_; }
+  void set_line_color(Color val) { this->line_color_ = val; }
+  const std::string get_name(void) { return name_; }
+  const HistoryData *get_tracedata() { return &data_; }
+
+ protected:
+  sensor::Sensor *sensor_{nullptr};
+  std::string name_{""};
+  uint8_t line_thickness_{3};
+  enum LineType line_type_ { LINE_TYPE_SOLID };
+  Color line_color_{COLOR_ON};
+  HistoryData data_;
+
+  friend Graph;
+  friend GraphLegend;
+};
+
+class Graph : public Component {
+ public:
+  void draw(display::DisplayBuffer *buff, uint16_t x_offset, uint16_t y_offset, Color color);
+  void draw_legend(display::DisplayBuffer *buff, uint16_t x_offset, uint16_t y_offset, Color color);
+
+  void setup() override;
+  float get_setup_priority() const override { return setup_priority::PROCESSOR; }
+  void dump_config() override;
+
+  void set_duration(uint32_t duration) { duration_ = duration; }
+  void set_width(uint32_t width) { width_ = width; }
+  void set_height(uint32_t height) { height_ = height; }
+  void set_min_value(float val) { this->min_value_ = val; }
+  void set_max_value(float val) { this->max_value_ = val; }
+  void set_min_range(float val) { this->min_range_ = val; }
+  void set_max_range(float val) { this->max_range_ = val; }
+  void set_grid_x(float val) { this->gridspacing_x_ = val; }
+  void set_grid_y(float val) { this->gridspacing_y_ = val; }
+  void set_border(bool val) { this->border_ = val; }
+  void add_trace(GraphTrace *trace) {
+    trace->init(this);
+    traces_.push_back(trace);
+  }
```

Comment:
Traces are also initialized in `setup()`. To me it _seems_ harmless to initialize them twice, but it's cleaner to not do it:
```suggestion
  void add_trace(GraphTrace *trace) { traces_.push_back(trace); }
```